automata/pda/pda.py

```diff
@@ -24,6 +47,313 @@ class PDA(Automaton, metaclass=abc.ABCMeta):
     initial_stack_symbol: str
     acceptance_mode: PDAAcceptanceModeT
 
+    @staticmethod
+    def _get_edge_name(
+        input_symbol: str = "", stack_top_symbol: str = "", stack_push: str = ""
+    ) -> str:
+        return (
+            ("ε" if input_symbol == "" else str(input_symbol))
+            + ", "
+            + ("ε" if stack_top_symbol == "" else str(stack_top_symbol))
+            + " | "
+            + ("ε" if stack_push == "" else str(stack_push))
+        )
+
+    @staticmethod
+    def _get_symbol_configuration(
+        from_state: PDAConfiguration, to_state: PDAConfiguration
+    ) -> Tuple[str, str, str]:
+        if (
+            from_state.remaining_input == to_state.remaining_input
+            or len(from_state.remaining_input) == 0
+        ):
+            input_symbol = ""
+        else:
+            input_symbol = from_state.remaining_input[0]
+
+        stack_top_symbol = from_state.stack.top()
+
+        if len(from_state.stack) == len(to_state.stack) + 1:
+            stack_push_symbols = ""
+        else:
+            stack_push_symbols = "".join(to_state.stack[len(from_state.stack) - 1 :])
+
+        return (input_symbol, stack_top_symbol, stack_push_symbols[::-1])
+
+    @abc.abstractmethod
+    def iter_transitions(
+        self,
+    ) -> Generator[Tuple[PDAStateT, PDAStateT, Tuple[str, str, str]], None, None]:
+        """
+        Iterate over all transitions in the automaton.
+        Each transition is a tuple of the form
+        (from_state, to_state, (input_symbol, stack_top_symbol, stack_push_symbols))
+        """
+
+    def show_diagram(
+        self,
+        input_str: Optional[str] = None,
+        with_machine: bool = True,
```

Comment:
Should add an error if both `with_machine` and `with_stack` are both false (since this almost certainly means a mistake in client code).

Reply:
Yes, it makes sense. I will update.